app/views/Project/Tagging/Sources/BulkUpload/index.tsx

```diff
@@ -1,65 +1,339 @@
+        [formError],
+    );
+
+    const {
+        setValue: onLeadChange,
+    } = useFormArray<'leads', PartialLeadType>('leads', setFormFieldValue);
+
+    const [
+        bulkCreateLeads,
+        // { loading: bulkCreateLeadsPending },
+    ] = useMutation<BulkCreateLeadsMutation, BulkCreateLeadsMutationVariables>(
+        BULK_CREATE_LEADS,
+        {
+            onCompleted: (response) => {
+                const leadBulk = response.project?.leadBulk;
+                if (!leadBulk) {
+                    return;
+                }
+                const {
+                    errors,
+                    result,
+                } = leadBulk;
+
+                if (errors) {
+                    const leadsErrors = errors?.map((item, index) => {
+                        if (isNotDefined(item)) {
+                            return undefined;
+                        }
+                        const clientId = leadClientIds?.[index];
+                        if (isNotDefined(clientId)) {
+                            return undefined;
+                        }
+
+                        return {
+                            clientId,
+                            error: transformToFormError(item),
+                        };
+                    }).filter(isDefined) ?? [];
+                    const leadsErrorMapping = listToMap(
+                        leadsErrors,
+                        (item) => item.clientId,
+                        (item) => item.error,
+                    );
+                    setFormError((oldError) => {
+                        const err = getErrorObject(oldError);
+                        return {
+                            ...err,
+                            leads: {
+                                ...getErrorObject(err?.leads),
+                                ...leadsErrorMapping,
+                            },
+                        };
+                    });
+                }
+                if (result) {
+                    const uploadedLeads = result?.map((item, index) => {
+                        if (isNotDefined(item)) {
+                            return undefined;
+                        }
+                        const clientId = leadClientIds?.[index];
+                        if (isNotDefined(clientId)) {
+                            return undefined;
+                        }
+
+                        return clientId;
+                    }).filter(isDefined) ?? [];
+                    if (uploadedLeads.length > 0) {
+                        alert.show(
+                            `${uploadedLeads.length} leads were successfully added!`,
+                            { variant: 'success' },
+                        );
+                        setLeadsAdded(true);
+                        setFormFieldValue((oldValues) => (
+                            oldValues?.filter((lead) => !uploadedLeads.includes(lead.clientId))
+                        ), 'leads');
```

Comment:
Don't we need to cleanup the attachments in memory?

Reply:
Also, this will clear out all of the leads that were errored (from the server or from the form)